### PR TITLE
Add support for FORGET pdu

### PR DESCRIPTION
**Pull request by GrayTShirt**
Looking for a little feedback

### CODE REVIEW

File: t/broadcast [hunk 1]
```diff
@@ -27,7 +27,7 @@ keysfile   ${ROOT}/var/keysdb
 max.events 5
 
 type :default {
-  freshness 3
+  freshness 10
```
Comment (jhunt) [inline t/broadcast]: Can this go into a separate commit?

File: t/forget [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+EOF
+
+sleep 1
+string_is "$(echo "FORGET|65535|test|0" | zdealer ${ZTK_OPTS} -c ${CONTROLLER})" \
+"OK"
+
+kill -TERM ${SUBSCRIBER_PID}
+kill -TERM ${BOLO_PID}
+
+exit 0
```
Comment (jhunt) [inline t/forget]: Should this test that the things got forgotten?

File: src/core.c [hunk 3]
```diff
@@ -450,6 +452,17 @@ int core_scheduler_thread(void *zmq, int interval) /* {{{ */
 
 /*************************************************************************/
 
+static int payload_is(uint16_t payload, uint16_t type) /* {{{ */
```
Comment (jhunt) [inline src/core.c]: I think this can be an & macro, instead of a branched function

File: src/bolo.h [hunk 4]
```diff
@@ -49,6 +49,15 @@
...
 #define PENDING  4
 
+#define P_STATE     0x0001
+#define P_COUNTER   0x0002
+#define P_RATE      0x0004
+#define P_SAMPLE    0x0008
+#define P_EVENT     0x0010
+#define P_FACT      0x0020
+#define P_RESERVED  0xFF70
+#define P_all       0xFFFF
```
Comment (jhunt) [inline src/bolo.h]: Can we make this P_ALL?

File: src/core.c [hunk 5]
```diff
@@ -734,6 +747,77 @@ static int _kernel_reactor(void *socket, pdu_t *pdu, void *_) /* {{{ */
...
+		if (_pdu_is(pdu, "FORGET", 4, 4)) {
+			const char *re_err;
+			char *pattern, *s;
+			int   re_off;
+			pcre *re;
+
+			s = pdu_string(pdu, 1); uint16_t payload   = strtoul(s, NULL, 10); free(s);
+			/* TODO ingore flag support
+			 *   s = pdu_string(pdu, 3); uint8_t  ignore    = strtoul(s, NULL, 10); free(s);
+			 */
```
Comment (jhunt) [inline src/core.c]: We should implement ignore

File: src/core.c [hunk 6]
```diff
@@ -51,6 +51,9 @@ typedef struct {
 #define winend(x, t)   (winstart((x), (t)) + (x)->window->time)
 #define max(a,b) ((a) > (b) ? (a) : (b))
 #define min(a,b) ((a) < (b) ? (a) : (b))
+#define payload_is(payload, type) \
+	(((payload) == PAYLOAD_ALL) ? 1 : ((payload) == (type)) ? 1 : 0)
```
Comment (jhunt) [inline src/core.c]: what about `((payload) & (type) > 0)`?

File: src/core.c [hunk 7]
```diff
@@ -734,6 +749,108 @@ static int _kernel_reactor(void *socket, pdu_t *pdu, void *_) /* {{{ */
...
 		/* }}} */
+		/* [ FORGET | type | pattern | ig ] {{{ */
+		if (_pdu_is(pdu, "FORGET", 4, 4)) {
+			const char *re_err;
+			char *pattern, *s;
+			int   re_off;
+			pcre *re;
+
+			s = pdu_string(pdu, 1); uint16_t payload   = strtoul(s, NULL, 10); free(s);
+			/* TODO ingore flag support */
```
Comment (jhunt) [inline src/core.c]: I don't think this is a TODO anymore...

File: t/lib [hunk 8]
```diff
@@ -190,3 +190,24 @@ string_like() {
   fi
   tdiag "ok ${msg}"
 }
+
+###############################################################################
+#
+# string_notlike - assert that two strings do not match via regex
+# USAGE: string_like ${actual} ${unexpected} \
```
Comment (jhunt) [inline t/lib]: Update the usage comment

**Comment by jhunt** [comment]
Overall I like it, with a few concerns:

1. BINF versioning
2. FORGET+broadcast semantics

For (1), do we need to bump the version on the binf file format?  Looks like we're adding a new attribute to each record type (ignored-or-not).  If old bolos will still parse the savedb, (because of record-length or something I'm forgetting right now), then no worries, otherwise we need to bump that version and figure out how to workaround older savedbs.

As for (2), I know the RFC doesn't spell out the FORGET+broadcast semantics (sorry, working on it) but my original thought was that we would tag each data point as "now ignored, please broadcast final" and then when it came time to do our broadcasts, we would broadcast for all currently live data points (as before) and all recently ignored data points (setting some other internal state to avoid repeat of that behavior).

Obviously, the `BROADCAST` PDU would have to be restructured to include a frame indicating if this was a FINAL broadcast or not, which will of course affect legacy subscribers.  Optionally, we can implement a new BROADCAST-like PDU called `FINAL` that we use to indicate that the broadcast is over.

In the future, we may want to extend the behavior of `FINAL` PDUs to have a configurable "window" during which to rebroadcast the `FINAL` message, so that subscribers have a period of time they can be out of communication and still receive the final notificaiton.

**Comment by GrayTShirt** [comment]
* Binf - yes an oversight on my behalf. It should be versioned, and no I new Core versions will be backwards compatible. I’m fine blowing away my savedb one last time.
* Broadcasting-FORGET - hmm this seems to tee into REBROADCASTING and autoscaling monitoring. Windowing seems to be a good idea, and FINAL would be a “jumbo” PDU, effectively a REBROADCAST?

**Comment by jhunt** [comment]
I was thinking something like this:

```
[STATE | old-server.legacy.foad | ... ]
[FORGET | STATE | old-server.legacy.foad]
[SAMPLE | cpu.old-server.legacy.foad | ... ]
[FORGET | SAMPLE | cpu.old-server.legacy.foad]

... time passes ...

[FORGET | STATE | old-server.legacy.foad]
[FORGET | SAMPLE | cpu.old-server.legacy.foad]
```

**Comment by GrayTShirt** [comment]
Updated and tested the above FORGET rebroadcast

**Comment by jhunt** [comment]
For now, remove the FORGET broadcast.  We can revisit the subscriber tie-in on the autoscaling thread.

Incidentally, how does an operator kick bolo to get it to forget something?  Should we go ahead and implement this is one of the CLI tools?

**Comment by GrayTShirt** [comment]
Should I comment out the broadcast code or just remove it?

As for the commands, I was going to implement in the perl app via wrelay. I can write some cli tooling if you like.

**Comment by jhunt** [comment]
Remove the broadcast code and add some CLI tooling.  Nobody here can see the Perl app :smile:

**Comment by GrayTShirt** [comment]
Updated with a documented forget utility (bolo_forget) :)

File: src/send.c [hunk 9]
```diff
@@ -19,6 +19,22 @@
...
+	char    *regex   = argv[1];
+
+	pdu_t *pdu = pdu_make("FORGET", 0);
+	pdu_extendf(pdu, "%u", payload);
+	pdu_extendf(pdu, "%s", regex);
+	pdu_extendf(pdu, "%u", ignore);
+
+	return pdu;
+}
+
```
Comment (jhunt) [inline src/send.c]: Where is `parse_forget_pdu` called, and why are we coercing `char*` to uint16_t / uint8_t

**Comment by GrayTShirt** [comment]
dropped

File: src/forget.c [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+		return 3;
+	}
+	if (vx_vzmq_connect(z, OPTIONS.endpoint) != 0) {
+		fprintf(stderr, "failed to connect to %s\n", OPTIONS.endpoint);
+		return 3;
+	}
+
+	pdu_t *p = forget_pdu(OPTIONS.payload, OPTIONS.pattern, OPTIONS.ignore);
+	if (pdu_send_and_free(p, z) != 0) {
+		fprintf(stderr, "failed to send [SET.KEYS] PDU to %s; command aborted\n", OPTIONS.endpoint);
```
Comment (jhunt) [inline src/forget.c]: Probably need to change this from `SET.KEYS` to `FORGET`
